CareKitStore/CareKitStore/CoreData/Synchronization/OCKStore+Synchronization.swift

```diff
@@ -215,8 +215,13 @@ extension OCKStore: OCKRemoteSynchronizationDelegate {
                                 self.context.knowledgeVector.increment(
                                     clockFor: self.context.clockID
                                 )
+                                
+                                // 8. Lock in the changes. If this fails, all
+                                //    merged changes will be rolled back and
+                                //    we'll need to try again later.
+                                try self.context.save()
```

Comment:
This bug wasn't easy to find, but was introduced with the changes in #553. The is is the last increment to the knowledge vector is never saved in context, causing data to be lost when saving OCK objects using a CareKit card or other view because it doesn't have the most up-to-date clock value.

The current test cases don't catch this issue because they are testing the stores directly as oppose to using the CareKit cards to save the outcomes.

The example case to replicate the issue is below:

1. Setup `Device1` to connect to a `OCKRemoteServer`
2. Add tasks
3. Save an outcome using a CareKit card and sync `Device1` to the `OCKRemoteServer`
4. Bring a new `Device2` online and sync it to the `OCKRemoteServer`. `Device2` will now reflect everything on `Device1`
5. Delete the saved outcome using a CareKit card on `Device2`. This will trigger a pullRevision, but notice during pushRevisions no new revision records will be found because the CareKit didn't have the correct Knowledge Vector clock value because it wasn't saved in context. Causing `Device1` to never receive the deleted outcome, forcing the devices out-of-sync

Reply:
This can also be done after push revisions is complete, but I'm not sure if it's actually better and requires extra code:

```swift
// 4. Detect and resolve any conflicts that exist. The store
                    //    is a CRDT, so there aren't any conflicts at the data
                    //    layer, but we want to resolve concurrent changes.
                    self.resolveConflicts { error in

                        self.context.perform {
                            do {
                                // 5. Lock in the changes. If this fails, all
                                //    merged changes will be rolled back and
                                //    we'll need to try again later.
                                try self.context.save()

                                // 6. Local revisions will now include any patches
                                //    that were applied during conflict resolution.
                                let localKnowledge = self.context.knowledgeVector

                                let localRevisions = try self.computeRevisions(
                                    since: remoteKnowledge
                                )

                                // 7. Bump knowledge vector to indicate that any
                                //    objects created beyond this point in time
                                //    are considered unknown to the peer.
                                self.context.knowledgeVector.increment(
                                    clockFor: self.context.clockID
                                )

                                // 8. Push conflict resolutions + local changes to remote
                                remote.pushRevisions(
                                    deviceRevisions: localRevisions,
                                    deviceKnowledge: localKnowledge) { error in

                                    if let error = error {
                                        os_log("Failed to push revision. %{private}@",
                                               log: .store, type: .error, error as NSError)
                                    }
                                    
                                    self.context.perform {
                                        do {
                                            // 9. Lock in the changes. If this fails, all
                                            //    merged changes will be rolled back and
                                            //    we'll need to try again later.
                                            try self.context.save()
                                        } catch {
                                            self.context.rollback()
                                            completion(error)
                                            return
                                        }
                                        // 10. The sync is still considered successful
                                        //    even if the remote doesn't accept the
                                        //    push. The next time we sync with it, it
                                        //    will still have the same knowledge
                                        //    vector, which means the proper diff will
                                        //    be generated.
                                        completion(nil)
                                    }
                                }
                            } catch {
                                self.context.rollback()
                                completion(error)
                            }
                        }
                    }
```

Reply:
Nice catch!